File: clients/typescript/src/debugging/debug-state.ts
import { Camera } from "./camera";
import { Vec2 } from "../model/vec2";
import { Stream } from "../stream";

/**
 * Renderer's state
 */
export class DebugState {
    /**
     * Pressed keys
     */
    pressedKeys: Array<string>
    /**
     * Cursor's position in game coordinates
     */
    cursorWorldPosition: Vec2
    /**
     * Id of unit which is followed by the camera, or None
     */
    lockedUnit: number | null
    /**
     * Current camera state
     */
    camera: Camera

    constructor(pressedKeys: Array<string>, cursorWorldPosition: Vec2, lockedUnit: number | null, camera: Camera) {
        this.pressedKeys = pressedKeys;
        this.cursorWorldPosition = cursorWorldPosition;
        this.lockedUnit = lockedUnit;
        this.camera = camera;
    }

    /**
     * Read DebugState from input stream
     */
    static async readFrom(stream: Stream): Promise<DebugState> {
        let pressedKeys;
        pressedKeys = [];
        for (let pressedKeysCount = await stream.readInt(); pressedKeysCount > 0; pressedKeysCount--) {
            let pressedKeysElement;
            pressedKeysElement = await stream.readString();
            pressedKeys.push(pressedKeysElement);
        }
        let cursorWorldPosition;
        cursorWorldPosition = await Vec2.readFrom(stream);
        let lockedUnit;
        if (await stream.readBool()) {
            lockedUnit = await stream.readInt();
        } else {
            lockedUnit = null;
        }
        let camera;
        camera = await Camera.readFrom(stream);
        return new DebugState(pressedKeys, cursorWorldPosition, lockedUnit, camera)
    }

    /**
     * Write DebugState to output stream
     */
    async writeTo(stream: Stream) {
        let pressedKeys = this.pressedKeys;
        await stream.writeInt(pressedKeys.length);
        for (let pressedKeysElement of pressedKeys) {
            await stream.writeString(pressedKeysElement);
        }
        let cursorWorldPosition = this.cursorWorldPosition;
        await cursorWorldPosition.writeTo(stream);
        let lockedUnit = this.lockedUnit;
        if (lockedUnit === null) {
            await stream.writeBool(false);
        } else {
            await stream.writeBool(true);
            await stream.writeInt(lockedUnit);
        }
        let camera = this.camera;
        await camera.writeTo(stream);
    }
}

File: clients/ruby/debugging/debug_state.rb
require './debugging/camera'
require './model/vec2'

module Debugging

# Renderer's state
class DebugState
    # Pressed keys
    attr_accessor :pressed_keys
    # Cursor's position in game coordinates
    attr_accessor :cursor_world_position
    # Id of unit which is followed by the camera, or None
    attr_accessor :locked_unit
    # Current camera state
    attr_accessor :camera

    def initialize(pressed_keys, cursor_world_position, locked_unit, camera)
        @pressed_keys = pressed_keys
        @cursor_world_position = cursor_world_position
        @locked_unit = locked_unit
        @camera = camera
    end

    # Read DebugState from input stream
    def self.read_from(stream)
        pressed_keys = []
        stream.read_int().times do |_|
            pressed_keys_element = stream.read_string()
            pressed_keys.push(pressed_keys_element)
        end
        cursor_world_position = Model::Vec2.read_from(stream)
        if stream.read_bool()
            locked_unit = stream.read_int()
        else
            locked_unit = nil
        end
        camera = Debugging::Camera.read_from(stream)
        DebugState.new(pressed_keys, cursor_world_position, locked_unit, camera)
    end

    # Write DebugState to output stream
    def write_to(stream)
        stream.write_int(@pressed_keys.length())
        @pressed_keys.each do |pressed_keys_element|
            stream.write_string(pressed_keys_element)
        end
        @cursor_world_position.write_to(stream)
        if @locked_unit.nil?
            stream.write_bool(false)
        else
            stream.write_bool(true)
            stream.write_int(@locked_unit)
        end
        @camera.write_to(stream)
    end

    def to_s
        string_result = "DebugState { "
        string_result += "pressed_keys: "
        string_result += "[ "
        pressed_keys_index = 0
        @pressed_keys.each do |pressed_keys_element|
            if pressed_keys_index != 0
                string_result += ", "
            end
            string_result += pressed_keys_element.dump
            pressed_keys_index += 1
        end
        string_result += " ]"
        string_result += ", "
        string_result += "cursor_world_position: "
        string_result += @cursor_world_position.to_s
        string_result += ", "
        string_result += "locked_unit: "
        if @locked_unit.nil?
            string_result += "nil"
        else
            string_result += @locked_unit.to_s
        end
        string_result += ", "
        string_result += "camera: "
        string_result += @camera.to_s
        string_result += " }"
        string_result
    end

    def to_str
        to_s
    end
end

end